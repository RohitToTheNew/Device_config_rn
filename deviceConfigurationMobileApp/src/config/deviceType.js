const deviceType = {
  ms700: 'MS-700',
  infoView: 'InfoviewPlayer',
  cza1300:'CZA1300'
};

export default deviceType;
