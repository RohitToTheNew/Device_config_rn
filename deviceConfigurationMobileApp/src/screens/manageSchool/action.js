const identifire = 'manage-school-log: ';
import { Alert } from 'react-native';
import RNFS from 'react-native-fs';
import Mailer from 'react-native-mail';
import Utils from '../../utils';
import { translate } from '../../translations/translationHelper';

/**
 * function to get saved Schools data from local storage
 */
function getSavedData() {
  const result = RNFS.readFile(RNFS.DocumentDirectoryPath + '/test.txt', 'utf8');
  return result;
}

/**
 * function for formatting the rooms data to desired CSV format
 * @param {object} parms object containing the data of rooms
 * @returns formatted CSV data
 */
function getDataInDesiredFormate(parms) {
  let CSVData = [];
  parms.forEach(room => {
    for (const key in room) {
      if (key === 'devices') {
        room[key].forEach(roomData => {
          let obj = {};
          obj.driverName = roomData.deviceType;
          obj.deviceName = roomData.deviceName;
          obj.ip = roomData.ipAddress || '';
          obj.mac = roomData.macAddress;
          obj.roomName = room['title'];
          CSVData.push(obj);
        });
      }
    }
  });
  return CSVData;
}

/**
 * function to save schools list to local storage
 * @param {object} schools object containing schools list
 * @returns 
 */
const writeDataTolocalStorage = async schools => {
  const data = JSON.stringify(schools);
  const path = RNFS.DocumentDirectoryPath + '/test.txt';
  try {
    await RNFS.writeFile(path, data, 'utf8');
    return `FILE WRITTEN!`;
  } catch (err) {
    Utils.Log(Utils.logType.error, 'error in writeDataTolocalStorage', err)
    return `Error while writing`;
  }
};

/**
 * function to write schools data to local storage
 * @param {object} schools object containing data of schools
 * @param {string} fileName name of file containing data of schools
 * @returns error(if any) during wiriting to local storage OR success message if writing is successful
 */

const writeCSVDataTolocalStorage = async (schools, fileName) => {
  const resultCSV = convertToCSV(schools);
  const pathToWrite = `${RNFS.DocumentDirectoryPath}/${fileName}.csv`;
  try {
    await RNFS.writeFile(pathToWrite, resultCSV, 'utf8');
    return `FILE WRITTEN!`;
  } catch (err) {
    Utils.showToast(translate('errorWhileWriting.'));
    return `Error while writing`;
  }
};
/**
 * function to attach generated CSV in email and open email client
 * @param {string} fileName name of file containing data of schools
 */
function sendCSVAsMail(schoolName, fileName) {
  try {
    Mailer.mail(
      {
        subject: `${schoolName} Devices`,
        isHTML: true,
        attachments: [
          {
            type: 'csv',
            path: `${RNFS.DocumentDirectoryPath}/${fileName}.csv`, // The absolute path of the file from which to read data.
          },
        ],
      },

      (error, event) => {
        Alert.alert(
          error,
          event,
          [
            {
              text: 'Ok',
              onPress: () => { },
            },
            {
              text: 'Cancel',
              onPress: () => { },
            },
          ],
          { cancelable: true },
        );
      },
    );
  } catch (error) {
    Utils.Log(Utils.logType.error, 'error in sendCSVAsMail', error);
  }
}

/**
 * function to handle the 'Export CSV' action for school
 * @param {object} parms object containig data of school
 */
async function handleExportCSV(parms) {
  try {
    const schoolName = parms?.schoolTitle;

    let month = new Date().getMonth() + 1;
    let date = new Date().getDate();
    const year = new Date().getFullYear();

    if (date < 10) date = '0' + date;
    if (month < 10) month = '0' + month;

    const hour = new Date().getHours();
    const minute = new Date().getMinutes();
    const second = new Date().getSeconds();

    const currentDate = `${month}-${date}-${year}`;
    const currentTime = `${hour}H${minute}M${second}S`;

    const fileName = `${schoolName}_Devices_${currentDate}_${currentTime}`
      .toString()
      .trim();
    const CSVData = getDataInDesiredFormate(parms?.rooms);
    await writeCSVDataTolocalStorage(CSVData, fileName);
    sendCSVAsMail(schoolName, fileName);
  } catch (error) {
    Utils.Log(Utils.logType.error, 'handleExportCSV', error);
  }
}

/**
 * function to convert the object data to desired CSV format
 * @param {object} arr array of rooms details
 * @returns
 */
function convertToCSV(arr) {
  const array = [Object.keys(arr[0])].concat(arr);
  return array
    .map(it => {
      return Object.values(it).toString();
    })
    .join('\n');
}

/**
 * function to show alert when delete button is pressed
 * @param {string} schoolTitle title of school
 * @param {string} schoolId school id
 * @param {function} setSchoolTitle function to set school title
 * @param {function} setSchoolId function to set school id
 * @param {function} setPopupToDeleteDevice function to set popup to show
 */
function deleteSchoolAlert(
  schoolTitle,
  schoolId,
  setSchoolTitle,
  setSchoolId,
  setPopupToDeleteDevice,
) {
  setSchoolTitle(schoolTitle);
  setSchoolId(schoolId);
  setPopupToDeleteDevice(true);
}

/**
 * function to delete school
 * @param {string} schoolID school id
 * @param {object} savedData object containing saved schools data
 * @param {function} setSavedData function to set saved schools data after change
 * @param {*} setPopupToDeleteDevice function to set popup to hide
 */
async function deleteSchool(
  schoolID,
  savedData,
  setSavedData,
  setPopupToDeleteDevice,
) {
  try {
    const filteredData = savedData.filter(school => school?.id !== schoolID);
    await writeDataTolocalStorage(filteredData);
    setSavedData(filteredData);
    setPopupToDeleteDevice(false);
  } catch (error) {
    Utils.Log(Utils.logType.error, 'deleteSchool', error);
  }
}
export {
  getSavedData,
  getDataInDesiredFormate,
  writeCSVDataTolocalStorage,
  sendCSVAsMail,
  handleExportCSV,
  deleteSchoolAlert,
  deleteSchool,
  writeDataTolocalStorage,
};
