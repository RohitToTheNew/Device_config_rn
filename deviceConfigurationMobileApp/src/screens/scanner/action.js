const identifire = 'scanner-action log: ';
import { check, request, RESULTS, PERMISSIONS } from 'react-native-permissions';
import { RNCamera } from 'react-native-camera';
import utils from '../../utils';
import { translate } from '../../translations/translationHelper';
import RNFS from 'react-native-fs';

/**
 * On click Enter details manually and redirect user to Add device screen
 * @param {object} navigation it represents navigation object
 * @returns object
 */
function actionHandleManualDetails(navigation) {
  return () => {
    navigation.navigate('addDevice');
  };
}

/**
 * function to read local file data from storage
 * @returns file details read from storage
 */
const readFileFromLocalStorage = async () => {
  let result = RNFS.readFile(RNFS.DocumentDirectoryPath + '/test.txt', 'utf8');
  return result;
};

/**
 * function to check if the device details encoded into the QR code is having valid keys
 * @param {object} obj object containing the scanned data from QR code
 * @returns boolean indicating weather details are valid or not
 */
const isValidObject = (obj) =>
  obj?.serialNumber && obj?.serialNumber !== '' &&
  obj?.macAddress && obj?.macAddress !== '' &&
  obj?.deviceType && obj?.deviceType !== '';

/**
 * function to check if the device details encoded into the QR code is having valid keys
 * @param {object} obj object containing the scanned data from QR code
 * @returns boolean indicating weather details are valid or not
 */
const isValidDetails = (obj) =>
  obj?.serial && obj?.serial !== '' &&
  obj?.mac && obj?.mac !== '' &&
  obj?.device && obj?.device !== '';

/**
 * On successful scan
 * @param {object} navigation it represents navigation object
 * @param {function} setInvalidCode used to set invalid code boolean value
 * @param {object} e contains scan result
 */
function actionOnSuccess(navigation, setInvalidCode, e) {
  const typeToCheck = utils.isIOS ? 'org.iso.QRCode' : "QR_CODE"
  if (e.type === typeToCheck) {
    try {
      const scannedString = JSON.parse(e?.data);
      console.log('scanned string', scannedString)
      if (isValidObject(scannedString) || isValidDetails(scannedString)) {
        navigation.replace('addDevice', {
          mac: scannedString?.macAddress || scannedString?.mac,
          serialNumber: scannedString?.serialNumber || scannedString?.serial,
          deviceType: scannedString?.deviceType || scannedString?.device,
        });
      } else {
        setInvalidCode(true);
      }
    } catch (e) {
      setInvalidCode(true);
    }
  } else {
    setInvalidCode(true);
  }
}

/**
 * Click on close icon
 * @param {object} navigation it represents navigation object
 * @returns object
 */
function actionHandleClose(navigation) {
  return () => {
    navigation.goBack();
  };
}

/**
 * Click on flash icon
 * @param {boolean} isFlashOn used to identify if flash is on/off
 * @param {function} setFlash used to set flash on/off value
 * @param {function} setIsFlashOn used to set flash on value
 * @returns object
 */
function actionHandleFlash(isFlashOn, setFlash, setIsFlashOn) {
  return () => {
    if (isFlashOn) {
      setFlash(RNCamera.Constants.FlashMode.off);
      setIsFlashOn(false);
    } else {
      setFlash(RNCamera.Constants.FlashMode.torch);
      setIsFlashOn(true);
    }
  };
}

/**
 * To ask camera permission
 * @returns camera permission
 */
async function actionUseEffect() {

  try {
    const result = await check(PERMISSIONS.IOS.CAMERA);
    if (result !== RESULTS.GRANTED) {
      const askCameraPermission = await request(PERMISSIONS.IOS.CAMERA);
      return askCameraPermission;
    } else {
      utils.Log(utils.logType.log, identifire, 'Camera permisson granted');
    }
  } catch (Ex) {
    utils.showToast(translate('error'), translate('permissionsException'));
  }
}
export {
  actionHandleManualDetails,
  actionOnSuccess,
  actionHandleClose,
  actionUseEffect,
  actionHandleFlash,
};
