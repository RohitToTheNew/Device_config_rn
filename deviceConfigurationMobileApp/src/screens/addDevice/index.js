import React, { useState } from 'react';
import {
  View,
  TouchableOpacity,
  Text,
  Modal,
  FlatList,
  TextInput,
  Keyboard,
} from 'react-native';
import { Mixins, Colors } from '../../config/styles';
import { translate } from '../../translations/translationHelper';
import { CloseWondow, SearchIcon } from '../../config/imageConstants';
import DropDown from '../dropDown/index';
import styles from './styles';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';

import { Header, ErrorWindow } from '../../components';
import {
  schoolDropDown,
  roomDropDown,
  schoolListUpdate,
  roomListUpdate,
  searchBarClose,
  renderListItem,
  onChangeTextFunction,
  saveDeviceEnableCheck,
  searchFilter,
  onBackPressSaveDevice,
  navigateToDeviceListing,
  deviceTypeDropDown,
  updateDeviceType,
  isAnyDataChanged,
  deleteSchoolOrRoomWithoutDevice,
} from './action';
import utils from '../../utils';
import DeviceType from '../../config/deviceType'

export default function AddDevice({ route, navigation }) {
  let deviceTypeScanIsValid = false;
  let deviceTypeScanValue = '';
  if (route?.params?.deviceType !== undefined) {
    const deviceTypeByScan = route?.params?.deviceType.toLowerCase();
    if (deviceTypeByScan === 'ms700' || deviceTypeByScan === 'infoviewplayer') {
      deviceTypeScanIsValid = true;
      if (deviceTypeByScan === 'ms700') {
        deviceTypeScanValue = 'MS700';
      } else if (deviceTypeByScan === 'infoviewplayer') {
        deviceTypeScanValue = 'infoViewPlayer';
      }
    } else {
      deviceTypeScanIsValid = true;
      deviceTypeScanValue = deviceTypeByScan.toUpperCase()
    }
  }

  const macAddressRef = React.createRef();
  const serialNumberRef = React.createRef();
  const deviceNameRef = React.createRef();

  const [dismissPopup, setDismissPopup] = useState(false);
  const [schoolId, setSchoolId] = useState(0);
  const [roomId, setRoomId] = useState(0);

  let [dataSource, setDataSource] = useState([]);
  let [schoolData, setSchoolData] = useState([]);
  let [roomData, setRoomData] = useState([]);

  const [modalVisible, setModalVisible] = useState(false);
  const [search, setSearch] = useState('');

  const [selectedSchool, setSelectedSchool] = useState(
    translate('selectSchool'),
  );
  const [selectedRoom, setSelectedRoom] = useState(translate('selectRoom'));

  const [deviceTypeSelected, setDeviceTypeSelected] = useState(false);
  const [deviceName, setDeviceName] = useState('');

  const [macAddress, setMacAddress] = useState(
    route?.params?.mac === undefined ? '' : route?.params?.mac,
  );
  const [serialNumber, setSerialNumber] = useState(
    route?.params?.serialNumber === undefined
      ? ''
      : route?.params?.serialNumber,
  );
  const [deviceType, setDeviceType] = useState(
    deviceTypeScanIsValid ? deviceTypeScanValue : 'Select Device Type',
  );

  let [isRoomDropDown, setIsRoomDropDown] = useState(false);
  let [schoolSelectedFlag, setSchoolSelectedFlag] = useState(false);
  let [roomSelectedFlag, setRoomSelectedFlag] = useState(false);
  let [deviceTypeSelectedFlag, setDeviceTypeSelectedFlag] = useState(false);

  const ListItem = renderListItem(
    isRoomDropDown,
    setSelectedRoom,
    setModalVisible,
    setSearch,
    dataSource,
    setSchoolId,
    setRoomId,
    setSelectedSchool,
    setSchoolSelectedFlag,
    setRoomSelectedFlag,
    roomData,
    schoolId,
    setIsRoomDropDown,
    setDeviceTypeSelected,
    modalVisible,
    setRoomData,
    setDataSource,
    deviceTypeScanIsValid
  );
  const renderItem = rowData => {
    return rowData.item.isNewEntry ? (
      <TouchableOpacity
        style={styles.renderItemSelectStyle}
        onPress={() =>
          isRoomDropDown
            ? updateRoomList(rowData.item.title)
            : updateSchoolList(rowData.item.title)
        }>
        <Text
          style={
            styles.renderItemCreateStyle
          }>{`+ Create ${rowData.item.title}`}</Text>
      </TouchableOpacity>
    ) : (
      <>
        <ListItem data={rowData} />
      </>
    );
  };
  const renderDeviceType = item => (
    <TouchableOpacity
      style={styles.item}
      onPress={() =>
        updateDeviceType(
          item.item.deviceType,
          setDeviceType,
          modalVisible,
          setModalVisible,
          setDeviceTypeSelectedFlag,
        )
      }>
      <Text style={styles.title}>{item.item.deviceType}</Text>
    </TouchableOpacity>
  );

  const updateSchoolList = schoolListUpdate(
    schoolData,
    setModalVisible,
    setSearch,
    setDataSource,
    setSchoolData,
    dataSource,
    setSchoolId,
    setSelectedSchool,
    setSelectedRoom,
    setSchoolSelectedFlag,
    setIsRoomDropDown,
    modalVisible,
    setDeviceTypeSelected,
    setRoomData,
  );

  const updateRoomList = roomListUpdate(
    dataSource,
    roomData,
    schoolId,
    setRoomData,
    setRoomId,
    setDataSource,
    setSelectedRoom,
    setModalVisible,
    setSearch,
    setRoomSelectedFlag,
    setIsRoomDropDown,
    modalVisible,
    setDeviceTypeSelected,
    deviceTypeScanIsValid
  );

  const handleCloseSearchBar = searchBarClose(
    setModalVisible,
    modalVisible,
    setSearch,
  );

  const headerRightButtons = () => {
    return <View style={styles.headerButtonContainer}></View>;
  };

  const handleSchoolDropDown = schoolDropDown(
    setIsRoomDropDown,
    setDeviceTypeSelected,
    modalVisible,
    setModalVisible,
    setSearch,
    setDataSource,
    setSchoolData,
  );

  const handleRoomDropDown = roomDropDown(
    schoolId,
    setIsRoomDropDown,
    setDeviceTypeSelected,
    setModalVisible,
    modalVisible,
    setSearch,
    setRoomData,
    setDataSource,
  );

  const searchFilterFunction = searchFilter(
    isRoomDropDown,
    roomData,
    schoolData,
    setDataSource,
    setSearch,
  );

  const isSaveDeviceEnable = saveDeviceEnableCheck(
    macAddress,
    serialNumber,
    selectedSchool,
    selectedRoom,
    deviceName,
    deviceType,
  );

  const onChangeText = onChangeTextFunction(
    setMacAddress,
    setSerialNumber,
    setDeviceName,
  );

  const handleBackPress = () => {
    let data = {
      schoolTitle: selectedSchool,
      roomTitle: selectedRoom,
      deviceType: deviceType,
      deviceName: deviceName,
      macAddress: macAddress,
      serialNumber: serialNumber,
    };
    // deleteSchoolOrRoomWithoutDevice();
    if (isAnyDataChanged(data)) {
      setDismissPopup(true);
    } else {
      navigation.goBack();
    }
  };

  return (
    <View style={styles.containerStyle}>
      <Header
        showBackButton={true}
        navigation={navigation}
        headerTitle={translate('addDevice')}
        overrideBackPress={handleBackPress}
      />
      <KeyboardAwareScrollView
        keyboardShouldPersistTaps="handled"
        contentContainerStyle={{ marginTop: Mixins.scaleSize(15) }}
        style={{ flexGrow: 1 }}
        bounces={false}>
        <View style={styles.dropDownContainerStyle}>
          <DropDown
            identity={'selectSchool'}
            testIDProp="schoolDropDown"
            isBorderRequired={true}
            isFieldSelected={schoolSelectedFlag}
            heading={'Select School'}
            name={selectedSchool}
            isDropDown={true}
            isEditable={false}
            onPress={handleSchoolDropDown}
          />
          <DropDown
            identity={'selectRoom'}
            testIDProp="roomDropDown"
            isBorderRequired={true}
            isFieldSelected={roomSelectedFlag}
            heading={'Select Room'}
            name={selectedRoom}
            isDropDown={true}
            isEditable={false}
            showDisabled={schoolId !== 0 ? false : true}
            onPress={() => schoolId !== 0 && handleRoomDropDown()}
          />

          <DropDown
            identity={'deviceType'}
            testIDProp="deviceTypeProp"
            isBorderRequired={true}
            isFieldSelected={deviceTypeSelectedFlag}
            heading={'Select Device Type'}
            name={deviceType}
            isDropDown={true}
            isEditable={false}
            onPress={() =>
              !deviceTypeScanIsValid &&
              deviceTypeDropDown(
                setIsRoomDropDown,
                modalVisible,
                setModalVisible,
                setSearch,
                setDataSource,
                setDeviceTypeSelected,
              )
            }
          />
          <DropDown
            identity={'deviceName'}
            isFieldSelected={true}
            heading={'Device Name'}
            ref={deviceNameRef}
            name={deviceName}
            isDropDown={false}
            maxLength={45}
            isEditable={true}
            onChangeText={onChangeText}
          />
          <DropDown
            identity={'macAddress'}
            testIDProp="macProp"
            isFieldSelected={true}
            maxLength={17}
            heading={'MAC Address'}
            ref={macAddressRef}
            name={macAddress}
            isDropDown={false}
            isEditable={route?.params?.mac === undefined ? true : false}
            onChangeText={onChangeText}
          />
          <DropDown
            identity={'serialNumber'}
            testIDProp="serialProp"
            isFieldSelected={true}
            heading={'Serial Number'}
            ref={serialNumberRef}
            name={serialNumber}
            isDropDown={false}
            isEditable={
              route?.params?.serialNumber === undefined ? true : false
            }
            onChangeText={onChangeText}
          />
          <View style={styles.saveButtonView}>
            <TouchableOpacity
              style={
                isSaveDeviceEnable
                  ? styles.saveButtonTO
                  : styles.saveButtonTODisable
              }
              disabled={isSaveDeviceEnable ? false : true}
              onPress={() => {
                Keyboard.dismiss();
                let params = {
                  macAddress,
                  serialNumber,
                  schoolId,
                  roomId,
                  deviceName,
                  deviceType,
                  selectedSchool,
                  selectedRoom,
                  navigation,
                };
                navigateToDeviceListing(params);
              }}>
              <Text style={styles.saveButtonText}>Save Device</Text>
            </TouchableOpacity>
          </View>
        </View>
      </KeyboardAwareScrollView>
      <Modal
        testID="listModel"
        animationType="slide"
        transparent={true}
        visible={modalVisible}
        onRequestClose={() => {
          setModalVisible(!modalVisible);
          setSearch('');
        }}>
        <View style={styles.centeredView}>
          <View style={styles.modalView}>
            <View style={styles.closeIcon}>
              <Text testID="dropDownHeading" style={styles.selectSchoolText}>
                {deviceTypeSelected
                  ? 'Select Device Type'
                  : isRoomDropDown
                    ? translate('selectRoom')
                    : translate('selectSchool')}
              </Text>
              <TouchableOpacity
                onPress={handleCloseSearchBar}
                style={styles.closeIconStyle}>
                <CloseWondow
                  width={Mixins.scaleSize(20)}
                  height={Mixins.scaleSize(20)}
                />
              </TouchableOpacity>
            </View>
            {!deviceTypeSelected && (
              <View style={styles.searchStyle}>
                <View style={styles.searchIcon}>
                  <SearchIcon
                    width={Mixins.scaleSize(14)}
                    height={Mixins.scaleSize(14)}
                  />
                </View>
                <TextInput
                  onChangeText={text => searchFilterFunction(text)}
                  value={search}
                  underlineColorAndroid="transparent"
                  placeholder={translate('typeToSearchOrAdd')}
                  placeholderTextColor={Colors.COLOR_808284}
                  style={styles.textInputStyle}
                />
              </View>
            )}
            <FlatList
              data={dataSource}
              renderItem={
                isRoomDropDown
                  ? renderItem
                  : isRoomDropDown === false && deviceTypeSelected === false
                    ? renderItem
                    : renderDeviceType
              }
              keyExtractor={item => item.id}
            />
          </View>
        </View>
      </Modal>
      <Modal
        visible={dismissPopup}
        transparent={true}
        hideModal={() => { }}
        backdrop={true}>
        <ErrorWindow
          errorTitle={translate('doYouWantToSaveThemBeforeMovingToOtherScreen')}
          button1Title={translate('save')}
          button2Title={translate('leave')}
          button1Action={() =>
            onBackPressSaveDevice(
              macAddress,
              serialNumber,
              schoolId,
              roomId,
              deviceName,
              deviceType,
              selectedSchool,
              selectedRoom,
              navigation,
              dismissPopup,
              setDismissPopup,
            )
          }
          button2Action={() => navigation.goBack()}
        />
      </Modal>
    </View>
  );
}
