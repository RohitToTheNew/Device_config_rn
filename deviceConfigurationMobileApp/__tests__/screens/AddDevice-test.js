import React from 'react';
import AddDevice from '../../src/screens/addDevice/index';
import renderer from 'react-test-renderer';
import { fireEvent, render, screen, cleanup, act } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import { store } from '../../src/store/configureStore';
import {
  roomDropDown,
  updateSelectedSchool,
  roomListUpdate,
  writeDataTolocalStorage,
  onBackPressSaveDevice,
  isDuplicateMacAddress,
} from '../../src/screens/addDevice/action';

import Toast from 'react-native-toast-message';
import * as addDeviceAction from '../../src/screens/addDevice/action';
import RNFS from 'react-native-fs';
import deviceType from '../../src/config/deviceType';

// Use fake timers to avoid async timer issues
jest.useFakeTimers();

// Global cleanup after each test
afterEach(() => {
  cleanup();
  jest.clearAllTimers();
});

const props = {
  writeDataTolocalStorage: jest
    .fn()
    .mockImplementation(() => Promise.resolve()),
  route: { params: [] },
  navigation: { push: () => jest.fn(), goBack: () => jest.fn() },
};

const tree = renderer.create(<AddDevice {...props} />).toJSON();
const tree1 = render(<AddDevice {...props} />);

describe('on landing to add device with scan result containing data', () => {
  it('should render add device correctly', () => {
    expect(tree).toMatchSnapshot();
  });
});

describe('on landing to add device with scan result containing no data', () => {
  const tree = renderer.create(<AddDevice {...props} />).toJSON();
  it('should render add device correctly', () => {
    expect(tree).toMatchSnapshot();
  });
});

describe('on landing to add device with scan result containing data', () => {
  let props, navigation;
  beforeEach(() => {
    navigation = { push: () => jest.fn() };
    props = {
      route: {
        params: { mac: 'AA:BB:CC:DD:EE:FF', serialNumber: 'SerialNumber 1', deviceType: deviceType.cza1300 },
      },
      navigation: navigation,
    };
    render(
      <Provider store={store}>
        <AddDevice {...props} />
      </Provider>,
    );
  });
  it('should render addDevice with macAddress, Serial number and device type pre-filled', async () => {
    const macProp = await screen.findByTestId('macPropselectedValue');
    expect(macProp).toHaveTextContent('AA:BB:CC:DD:EE:FF');
    const serialProp = await screen.findByTestId('serialPropselectedValue');
    expect(serialProp).toHaveTextContent('SerialNumber 1');
    const deviceTypeProp = await screen.findByTestId('deviceTypeProp');
    expect(deviceTypeProp).toHaveTextContent('CZA1300')
  });
});

describe('Write file', () => {
  const mockData = [
    {
      id: 1,
      title: 'School 1',
      rooms: [
        {
          id: 1,
          title: 'Room 1 for School 1',
          devices: [
            {
              macAddress: 'macaddreeesssss12',
              serialNumber: 'serial number 2',
              deviceName: "Developer's device",
              deviceType: 'MS700',
              deviceID: 2,
            },
          ],
        },
      ],
    },
  ];

  test('should write a file', async () => {
    const result = await writeDataTolocalStorage(mockData);
    expect(result).toEqual('FILE WRITTEN!');
  });
  test('should throw an error while write a file', async () => {
    jest.spyOn(RNFS, 'writeFile').mockImplementation(() => {
      throw new Error('Error while writing');
    });

    const result = await writeDataTolocalStorage(mockData);
    expect(result).toEqual('Error while writing');
  });
});

describe('Room drop down', () => {
  test('on selection of roomDropDown when school is not selected it should show toast', () => {
    let schoolId = 0;
    const setIsRoomDropDown = () => { };
    const setModalVisible = () => { };
    const modalVisible = () => { };
    const setSearch = () => { };
    const setRoomData = () => { };
    const setDataSource = () => { };

    jest.resetAllMocks();
    roomDropDown(
      schoolId,
      setIsRoomDropDown,
      setModalVisible,
      modalVisible,
      setSearch,
      setRoomData,
      setDataSource,
    );
    const toastSpy = jest.spyOn(Toast, 'show');
    expect(toastSpy).toBeCalled();
  });
});

describe('School drop down', () => {
  test('It should create `New School` and select it', async () => {
    render(
      <Provider store={store}>
        <AddDevice {...props} />
      </Provider>,
    );
    fireEvent.press(screen.getByTestId('schoolDropDownTouchable'));
    fireEvent.changeText(screen.getByTestId('searchBar'), 'New School');
    expect(screen.getByTestId('rendeItemBtn')).toBeTruthy();
    expect(screen.getByTestId('listModel')).toBeTruthy();

    fireEvent.press(screen.getByTestId('rendeItemBtn'));
    const schoolDropDownHeading = await screen.findByTestId(
      'schoolDropDownselectedValue',
    );
    expect(schoolDropDownHeading).toHaveTextContent('New School');

    fireEvent.press(screen.getByTestId('roomDropDownTouchable'));
    fireEvent.changeText(screen.getByTestId('searchBar'), 'New Room');

    expect(screen.getByTestId('rendeItemBtn')).toBeTruthy();
    expect(screen.getByTestId('listModel')).toBeTruthy();

    fireEvent.press(screen.getByTestId('rendeItemBtn'));
    const roomDropDownHeading = await screen.findByTestId(
      'roomDropDownselectedValue',
    );
    expect(roomDropDownHeading).toHaveTextContent('New Room');
  });

  describe('Update Selected School', () => {
    test('on Selection of a school having School 1 as a title, Room title should be reset to default', async () => {
      render(
        <Provider store={store}>
          <AddDevice {...props} />
        </Provider>,
      );

      const dataSource = [
        {
          id: 1,
          title: 'School 1',
          rooms: [
            {
              id: 1,
              title: 'Room 1',
              devices: [
                {
                  macAddress: 'AAAAAAAAAAAAAAAAA',
                  serialNumber: 'serial number',
                },
              ],
            },
          ],
        },
      ];
      const title = 'School 1';
      const setSchoolId = () => { };
      setSelectedSchool = () => { };
      setSelectedRoom = () => { };
      setModalVisible = () => { };
      setSearch = () => { };
      setSchoolSelectedFlag = () => { };

      updateSelectedSchool(
        dataSource,
        title,
        setSchoolId,
        setSelectedSchool,
        setSelectedRoom,
        setModalVisible,
        setSearch,
        setSchoolSelectedFlag,
      );

      const roomDropDownHeading = await screen.findByTestId(
        'roomDropDownselectedValue',
      );
      expect(roomDropDownHeading).toHaveTextContent(translate('selectRoom'));
    });
  });

  describe('Mac Address input', () => {
    test('on input to mac address, it should set mac address field', async () => {
      const onChangeTextFnSPy = jest.spyOn(
        addDeviceAction,
        'onChangeTextFunction',
      );

      render(
        <Provider store={store}>
          <AddDevice {...props} />
        </Provider>,
      );

      fireEvent.press(screen.getByTestId('macProp'));
      fireEvent.changeText(screen.getByTestId('macProp'), 'abcdefghijkl');

      expect(onChangeTextFnSPy).toHaveBeenCalled();
    });
  });

  describe('Serial number input', () => {
    test('on input to serial number, it should set serial number field', async () => {
      const onChangeTextFnSPy = jest.spyOn(
        addDeviceAction,
        'onChangeTextFunction',
      );
      render(
        <Provider store={store}>
          <AddDevice {...props} />
        </Provider>,
      );

      fireEvent.press(screen.getByTestId('serialProp'));
      fireEvent.changeText(screen.getByTestId('serialProp'), 'serial number 1');
      expect(onChangeTextFnSPy).toHaveBeenCalled();
    });
  });

  describe('on back press', () => {
    test('if submitted with invalid macaddress, it should show toast', async () => {
      render(
        <Provider store={store}>
          <AddDevice {...props} />
        </Provider>,
      );

      const macAddress = '';
      const serialNumber = 'serial number 1';
      const schoolId = '1';
      const roomId = '1';
      const navigation = () => { };

      onBackPressSaveDevice(
        macAddress,
        serialNumber,
        schoolId,
        roomId,
        navigation,
      );

      const toastSpy = jest.spyOn(Toast, 'show');
      expect(toastSpy).toBeCalled();
    });
  });

  describe('on back press', () => {
    test('if submitted with duplicate macaddress, it should show toast for duplicate mac', async () => {
      render(
        <Provider store={store}>
          <AddDevice {...props} />
        </Provider>,
      );

      const macAddress = 'AB:CD:EF:GH:VV';
      const serialNumber = 'serial number 1';
      const schoolId = '1';
      const roomId = '1';
      const navigation = () => { };

      onBackPressSaveDevice(
        macAddress,
        serialNumber,
        schoolId,
        roomId,
        navigation,
      );

      const toastSpy = jest.spyOn(Toast, 'show');
      expect(toastSpy).toBeCalled();
    });
  });

  describe('duplicate mac function', () => {
    test('if macAddress is duplicate, it should return true', async () => {
      render(
        <Provider store={store}>
          <AddDevice {...props} />
        </Provider>,
      );

      const macAddress = 'AB:LW:EF:GH:IJ';
      const schoolId = 1;

      const result = await isDuplicateMacAddress(schoolId, macAddress);
      expect(result).toBeTruthy();
    });
  });
});

describe('RoomDropDown Selection', () => {
  test('if school is selected, it should fetch rooms list under selected school', () => {
    let schoolId = 1;
    const setIsRoomDropDown = () => { };
    const setModalVisible = () => { };
    const modalVisible = () => { };
    const setSearch = () => { };
    const setRoomData = () => { };
    const setDataSource = () => { };

    const readValueSpy = jest.spyOn(RNFS, 'readFile');
    roomDropDown(
      schoolId,
      setIsRoomDropDown,
      setModalVisible,
      modalVisible,
      setSearch,
      setRoomData,
      setDataSource,
    );
    expect(readValueSpy).toBeCalled();
  });

  test('on selection of roomDropDown when school is selected and invoke getRoom', () => {
    let schoolId = 1;
    const setIsRoomDropDown = () => { };
    const setModalVisible = () => { };
    const modalVisible = () => { };
    const setSearch = () => { };
    const setRoomData = () => { };
    const setDataSource = () => { };

    const readValueSpy = jest.spyOn(RNFS, 'readFile').mockImplementation(() => {
      throw new Error('cannot read data from localStorage.');
    });
    roomDropDown(
      schoolId,
      setIsRoomDropDown,
      setModalVisible,
      modalVisible,
      setSearch,
      setRoomData,
      setDataSource,
    );
    expect(readValueSpy).toBeCalled();
  });

  test('it should show toast if school is not selected', () => {
    let schoolId = 0;
    const setIsRoomDropDown = () => { };
    const setModalVisible = () => { };
    const modalVisible = () => { };
    const setSearch = () => { };
    const setRoomData = () => { };
    const setDataSource = () => { };

    jest.resetAllMocks();
    roomDropDown(
      schoolId,
      setIsRoomDropDown,
      setModalVisible,
      modalVisible,
      setSearch,
      setRoomData,
      setDataSource,
    );
    const toastSpy = jest.spyOn(Toast, 'show');
    expect(toastSpy).toBeCalled();
  });
});
