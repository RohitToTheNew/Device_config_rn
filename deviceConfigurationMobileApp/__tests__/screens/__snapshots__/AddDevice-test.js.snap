// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on landing to add device with scan result containing data should render add device correctly 1`] = `null`;

exports[`on landing to add device with scan result containing no data should render add device correctly 1`] = `null`;
